include/fastcdr/Cdr.h

```diff
@@ -733,23 +746,89 @@ class Cdr
     {
-        if (!is_multi_array_primitive(&array_t))
-        {
-            Cdr::state dheader_state {allocate_xcdrv2_dheader()};
+        serialize_array(array_t.data(), array_t.size());
+        return *this;
+    }
+
+    /*!
+     * @brief This function template serializes an array.
+     * @param array_t The array that will be serialized in the buffer.
+     * @return Reference to the eprosima::fastcdr::Cdr object.
+     * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
+     */
+    template<class _T, size_t _Size,
+            typename std::enable_if<
+                !static_is_multi_array_primitive<std::array<_T, _Size> const*>::value &&
+                is_complex_array_or_string<std::array<_T, _Size>>::value>::type* = nullptr>
+    Cdr& serialize(
+            const std::array<_T, _Size>& array_t)
+    {
+        Cdr::state dheader_state {allocate_xcdrv2_dheader()};
+
+        serialize_array(array_t.data(), array_t.size());
+
+        set_xcdrv2_dheader(dheader_state);
+
+        return *this;
+    }
+
+    /*!
+     * @brief This function template serializes an array.
+     * @param array_t The array that will be serialized in the buffer.
+     * @return Reference to the eprosima::fastcdr::Cdr object.
+     * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
+     */
+    template<class _T, size_t _Size,
+            typename std::enable_if<
+                !static_is_multi_array_primitive<std::array<_T, _Size> const*>::value &&
+                !is_complex_array_or_string<std::array<_T, _Size>>::value>::type* = nullptr>
+    Cdr& serialize(
+            const std::array<_T, _Size>& array_t)
+    {
+        Cdr::state dheader_state {allocate_xcdrv2_dheader()};
+
+        eprosima::fastcdr::serialize_array(*this, array_t.data(), array_t.size());
+
+        set_xcdrv2_dheader(dheader_state);
+
+        return *this;
+    }
+
+    /*!
+     * @brief This function template serializes a sequence of non-primitive.
+     * @param vector_t The sequence that will be serialized in the buffer.
+     * @return Reference to the eprosima::fastcdr::Cdr object.
+     * @exception exception::NotEnoughMemoryException This exception is thrown when trying to serialize a position that exceeds the internal memory size.
+     */
+    template<class _T, typename std::enable_if<
+                !std::is_enum<_T>::value &&
+                !std::is_arithmetic<_T>::value &&
+                !is_complex_array_or_string<std::vector<_T>>::value>::type* = nullptr>
+    Cdr& serialize(
+            const std::vector<_T>& vector_t)
+    {
+        Cdr::state dheader_state {allocate_xcdrv2_dheader()};
 
-            serialize_array(array_t.data(), array_t.size());
+        serialize(static_cast<int32_t>(vector_t.size()));
```

Comment:
```suggestion
        serialize(static_cast<uint32_t>(vector_t.size()));
```

Reply:
We should check that the size is not bigger than the maximum value of a `uint32_t`.
I propose having a `serialize_collection_length(size_t length)` method that throws an exception when `length > MAX_UINT32`, and then serializes `static_cast<uint32_t>(length)`.

Reply:
Note: this would imply changes in other methods unrelated to this PR. It might be a good idea to do that refactoring on a separate PR that we can backport.